### PR TITLE
Implement a monotonic cubic interpolator with at least C1 continuity

### PR DESCRIPTION
**Pull request by lindsayad**
Algorithm is based on Fritsch's and Carlson's 1980 paper "Monotone Piecewise Cubic Interpolation." In brief, the algorithm is:

1. From input data set, initialize derivatives using second order accurate differencing
2.  Check derivatives against a region of monotonicity. 
3. If outside monotonicity region, modify derivatives accordingly.

References issue #7158

### CODE REVIEW

**Comment by lindsayad** [comment]
Still to do: 
 - Add sample2ndDerivative
 - squash small commits

**Review by permcody — COMMENTED** [review]
This looks pretty good. I'll make a general comment about these utilities. I'd like to see them separated a little more from MOOSE. Some of them are, some of them aren't. LinearInterpolation is one of the better ones. It throws errors instead of calling MOOSE errors and is a stand alone utility. It even has tests in the unit suite. Some of these other utilities are linked into MOOSE (call mooseError() for instance) and only have full integration tests.

File: framework/include/utils/MonotoneCubicInterpolation.h [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+
+class MonotoneCubicInterpolation
+{
+public:
+  MonotoneCubicInterpolation();
+  MonotoneCubicInterpolation(const std::vector<Real> x, const std::vector<Real> y);
+
+  virtual ~MonotoneCubicInterpolation() = default;
+
+  virtual void setData(const std::vector<Real> x, const std::vector<Real> y);
```
Comment (jwpeterson) [inline framework/include/utils/MonotoneCubicInterpolation.h]: `x` and `y` should be passed by const reference.

File: framework/include/utils/MonotoneCubicInterpolation.h [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+  virtual Real sample(Real x) const;
+  virtual Real sampleDerivative(Real x) const;
+  virtual void dumpCSV(std::string filename, const std::vector<Real> xnew);
```
Comment (jwpeterson) [inline framework/include/utils/MonotoneCubicInterpolation.h]: `xnew` should be passed by const reference.

File: framework/include/utils/MonotoneCubicInterpolation.h [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  Real h1Prime(const Real & xhi, const Real & xlo, const Real & x) const;
+  Real h2Prime(const Real & xhi, const Real & xlo, const Real & x) const;
+  Real h3Prime(const Real & xhi, const Real & xlo, const Real & x) const;
+  Real h4Prime(const Real & xhi, const Real & xlo, const Real & x) const;
+
+  /// Interpolating cubic polynomial
+  virtual Real p(const Real & xhi, const Real & xlo, const Real & fhi, const Real & flo,
+                 const Real & dhi, const Real & dlo, const Real & x) const;
+  virtual Real pPrime(const Real & xhi, const Real & xlo, const Real & fhi, const Real & flo,
+                 const Real & dhi, const Real & dlo, const Real & x) const;
```
Comment (jwpeterson) [inline framework/include/utils/MonotoneCubicInterpolation.h]: Indentation does not line up with line above.

File: framework/include/utils/MonotoneCubicInterpolation.h [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+#include <vector>
+#include "libmesh/libmesh_common.h"
+using libMesh::Real;
+
+class MonotoneCubicInterpolation
+{
+public:
+  MonotoneCubicInterpolation();
+  MonotoneCubicInterpolation(const std::vector<Real> x, const std::vector<Real> y);
```
Comment (jwpeterson) [inline framework/include/utils/MonotoneCubicInterpolation.h]: `x` and `y` should be passed by const reference.

File: framework/src/utils/MonotoneCubicInterpolation.C [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+
+  checkMonotone();
+  if (_monotonic_e == "monotonic_not")
+    mooseError("Don't ask for a monotonic interpolation routine if your dependent variable data isn't monotonic.");
+}
+
+Real
+MonotoneCubicInterpolation::sign(Real x) const
+{
+  return x < 0 ? -1. : (x > 0 ? 1. : 0.);
```
Comment (jwpeterson) [inline framework/src/utils/MonotoneCubicInterpolation.C]: xz

File: framework/src/utils/MonotoneCubicInterpolation.C [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+    y_diff = _y[i+1] - _y[i];
+    if (s == 0)
+      s = sign(y_diff);
+    if (s * y_diff < 0)
+    {
+      _monotonic_e = "monotonic_not";
+      return;
+    }
+  }
+  s > 0 ? _monotonic_e = "monotonic_increase" :
```
Comment (jwpeterson) [inline framework/src/utils/MonotoneCubicInterpolation.C]: Please just use normal if-statements, it's much easier to read.

File: framework/src/utils/MonotoneCubicInterpolation.C [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+    (s < 0 ? _monotonic_e = "monotonic_decrease" :
+     _monotonic_e = "montonic_constant");
+}
+
+Real
+MonotoneCubicInterpolation::phi(const Real & t) const
+{
+  return 3. * std::pow(t, 2) - 2. * std::pow(t, 3);
```
Comment (jwpeterson) [inline framework/src/utils/MonotoneCubicInterpolation.C]: `t*t` and `t*t*t` are shorter and just as readable as `std::pow` calls IMO.

**Review by jwpeterson — DISMISSED** [review]

**Comment by lindsayad** [comment]
Hopefully I addressed your review @jwpeterson. @permcody I migrated my tests over to unit and removed all dependence of the interpolation utility on MOOSE.

**Comment by permcody** [comment]
Thanks @lindsayad, this looks much better!

**Review by permcody — APPROVED** [review]

**Comment by permcody** [comment]
@brianmoose - Thanks for fixing up the "developer activated logic". Please add @lindsayad to the approved group (Moose Collaborators?).

**Comment by dschwen** [comment]
Why does this use `double` all over the place?

**Comment by lindsayad** [comment]
Instead of `Real`? Based off @permcody's recommendation, I divorced this class completely from libmesh/MOOSE such that it should be able to be a stand-alone utility.

**Comment by dschwen** [comment]
I don't think that is a great idea. It is in the MOOSE framework directory. There is a (small) risk that we run into cases where Real != double. And in my opinion (and that might sound a bit harsh ;-) ) this _pollutes_ the MOOSE code in which we very consistently use Real (unless we are coupling to external libraries or fortran codes).

**Comment by lindsayad** [comment]
I don't take any offense. This is a difference of opinion that I don't care so much about. Whether the team wants it to be a completely stand-alone utility or whether it should maintain consistency with Real, it's only a `sed` command away. I'll abide by your opinion.

**Comment by lindsayad** [comment]
I will say that the `SplineInterpolation` utility used `double` until I came and re-wrote it in order to add the `BicubicSpline` and switched all occurrences with `Real` ;-)

**Comment by permcody** [comment]
I'll chime in. I'd actually prefer using `Real` too. I'm very happy that
you pulled this out of MOOSE as requested but it's possible to still
include the header with the typedef to use `Real`. It can be added back in
if there are people that would like to see it that way.

**Comment by lindsayad** [comment]
Alrighty, double swapped back to Real

**Comment by lindsayad** [comment]
Thanks @jwpeterson!

**Comment by jwpeterson** [comment]
No problem.  After merging, I realized there's not really any documentation for the new class.  Would you mind adding a Doxygen-style comment block to the top of the class, i.e.
```
/**
 * Blah blah blah
 */
```
which gives the reference to Fritsch and Carlson's paper, and document any parts of the public interface which users are supposed to call.

**Comment by lindsayad** [comment]
Yea, I'll get on it.

**Pull request merged**